dot/parachain/candidate-validation/candidate_validation.go

```diff
@@ -211,3 +232,110 @@ func validateFromChainState(runtimeInstance parachainruntime.RuntimeInstance, po
 
 	return &candidateCommitments, persistedValidationData, isValid, nil
 }
+
+// validateFromExhaustive validates a candidate parachain block with provided parameters
+func validateFromExhaustive(validationHost parachainruntime.ValidationHost,
+	persistedValidationData parachaintypes.PersistedValidationData,
+	validationCode parachaintypes.ValidationCode,
+	candidateReceipt parachaintypes.CandidateReceipt, pov parachaintypes.PoV) (
+	*ValidationResult, error) {
+
+	validationCodeHash := validationCode.Hash()
+	// basic checks
+	validationErr, internalErr := performBasicChecks(&candidateReceipt.Descriptor, persistedValidationData.MaxPovSize,
+		pov,
+		validationCodeHash)
+	if internalErr != nil {
+		return nil, fmt.Errorf("performing basic checks: %w", internalErr)
+	}
+
+	if validationErr != nil {
+		validationResult := &ValidationResult{
+			InvalidResult: validationErr,
+		}
+		return validationResult, nil //nolint: nilerr
+	}
+
+	validationParams := parachainruntime.ValidationParameters{
+		ParentHeadData:         persistedValidationData.ParentHead,
+		BlockData:              pov.BlockData,
+		RelayParentNumber:      persistedValidationData.RelayParentNumber,
+		RelayParentStorageRoot: persistedValidationData.RelayParentStorageRoot,
+	}
+
+	validationResult, err := validationHost.ValidateBlock(validationParams)
+	if err != nil {
+		return nil, fmt.Errorf("executing validate_block: %w", err)
+	}
+
+	validationResultHash, err := validationResult.HeadData.Hash()
+	if err != nil {
+		return nil, fmt.Errorf("hashing validation result: %w", err)
+	}
+
+	if validationResultHash != candidateReceipt.Descriptor.ParaHead {
+		ci := ParaHeadHashMismatch
+		return &ValidationResult{InvalidResult: &ci}, nil
+	} else {
+		outputs := parachaintypes.CandidateCommitments{
+			UpwardMessages:            validationResult.UpwardMessages,
+			HorizontalMessages:        validationResult.HorizontalMessages,
+			NewValidationCode:         validationResult.NewValidationCode,
+			HeadData:                  validationResult.HeadData,
+			ProcessedDownwardMessages: validationResult.ProcessedDownwardMessages,
+			HrmpWatermark:             validationResult.HrmpWatermark,
+		}
+
+		// if validation produced a new set of commitments, we treat the candidate as invalid
+		if candidateReceipt.CommitmentsHash != outputs.Hash() {
+			ci := CommitmentsHashMismatch
+			return &ValidationResult{InvalidResult: &ci}, nil
+		} else {
+			return &ValidationResult{
+				ValidResult: &ValidValidationResult{
+					CandidateCommitments:    outputs,
+					PersistedValidationData: persistedValidationData,
+				},
+			}, nil
+		}
+	}
```

Comment:
- Let's rename `output` to `candidateCommitments`.
- Also simplify the code by removing unnecessary `else` blocks. That would be more readable, IMO

```suggestion
	if validationResultHash != candidateReceipt.Descriptor.ParaHead {
		ci := ParaHeadHashMismatch
		return &ValidationResult{InvalidResult: &ci}, nil
	}

	candidateCommitments := parachaintypes.CandidateCommitments{
		UpwardMessages:            validationResult.UpwardMessages,
		HorizontalMessages:        validationResult.HorizontalMessages,
		NewValidationCode:         validationResult.NewValidationCode,
		HeadData:                  validationResult.HeadData,
		ProcessedDownwardMessages: validationResult.ProcessedDownwardMessages,
		HrmpWatermark:             validationResult.HrmpWatermark,
	}

	// if validation produced a new set of commitments, we treat the candidate as invalid
	if candidateReceipt.CommitmentsHash != candidateCommitments.Hash() {
		ci := CommitmentsHashMismatch
		return &ValidationResult{InvalidResult: &ci}, nil
	}

	return &ValidationResult{
		ValidResult: &ValidValidationResult{
			CandidateCommitments:    candidateCommitments,
			PersistedValidationData: persistedValidationData,
		},
	}, nil
```

Reply:
updated.